### PR TITLE
QuantumState init from_array added and state sampling functionality given

### PR DESCRIPTION
**Pull request by AlexisRalli**
Can initalize quantum state from an array directly now.

Can also sample state in computational basis

### CODE REVIEW

File: symmer/symplectic/base.py [hunk 1]
```diff
@@ -999,6 +998,64 @@ def to_sparse_matrix(self):
...
+
+        if not np.isclose(np.linalg.norm(statevector), 1):
+            warnings.warn(f'statevector is not normalized')
+
+        N = int(N)
+        non_zero = np.where(abs(statevector) >= threshold)[0]
+
+        # build binary states of non_zero terms
+        if N<64:
+            state_matrix = (((non_zero[:, None] & (1 << np.arange(N))[::-1])) > 0).astype(int)
```
Comment (AlexisRalli) [inline symmer/symplectic/base.py:1052]: fast state_matrix generation!

File: symmer/symplectic/base.py [hunk 2]
```diff
@@ -999,6 +998,64 @@ def to_sparse_matrix(self):
...
+        #TODO: could have change of operator added prior to sample
+
+        Args:
+            n_samples: how many bitstring samples to take
+
+        Returns:
+            self.state_matrix: matrix, where rows represent bitstrings
+            counter: number of times each row was sampled
+        """
+        counter = np.random.multinomial(n_samples, np.abs(self.state_op.coeff_vec)**2)
```
Comment (AlexisRalli) [inline symmer/symplectic/base.py:1014]: fast way to sample a state!

qiskit uses: https://github.com/Qiskit/qiskit-terra/blob/6abb8824e54ff7fccb589fe6a0f3912ebebd95d0/qiskit/quantum_info/states/quantum_state.py#L271

np unique is SLOW!

Comment (AlexisRalli) [reply symmer/symplectic/base.py:1014]: @TimWeaving look at commented out part above return statement... We could return as a qubit state, with coeffs = number of samples! What do you think?

Comment (TimWeaving) [reply symmer/symplectic/base.py]: Agreed on returning as QuantumState, with option to return raw counts or normalized state

**Review by TimWeaving — APPROVED** [review]
Stricter conditions on bra/ket state initialization mean less likely to make mistakes when converting arrays, i.e. column vector = ket, row vector = bra!

**Pull request merged**